Update for `gradle/actions@v3.0.0` release

bigdaz opened the pull request:
This PR updates the existing `gradle.yml` and `gradle-publish.yml` starter workflows for recent updates to `gradle/gradle-build-action`.

- Replace uses of `gradle/gradle-build-action` with `gradle/actions/setup-gradle`
- Split "Setup Gradle" step out of "Run Gradle" steps
- Add separate Job using `gradle/actions/dependency-submission` (flagged as optional)
- Bump version hashes to use `v3.0.0` of Gradle actions
- Bump to Node20-compatible versions of `actions/checkout` and `actions/setup-java`
- Bump to setup JDK 17

<!--
IMPORTANT:

This repository contains configuration for what users see when they click on the `Actions` tab and the setup page for Code Scanning.

It is not:
* A playground to try out scripts
* A place for you to create a workflow for your repository
-->

## Pre-requisites

- [ ] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [ ] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [ ] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [ ] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [ ] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [ ] Should specify least privileged [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [ ] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [ ] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/code-scanning).
- [ ] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [ ] `name`: Name of the Code Scanning integration.
  - [ ] `creator`: Name of the organization/user producing the Code Scanning integration.
  - [ ] `description`: Short description of the Code Scanning integration.
  - [ ] `categories`: Array of languages supported by the Code Scanning integration.
  - [ ] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [ ] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [ ] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [ ] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [ ] Automation and CI workflows cannot be dependent on a paid service or product.

jonjanego commented:
Summarizing our private slack conversation @bigdaz:

My biggest suggested change is to make the dependency submission job on by default, with some explanatory text in a comment. This will end up with the best quality dependency results for customers.

Could you also add some explanation in a comment as to the change to running `gradlew` instead of running `gradle-build-action` ?

Thank you!

bigdaz commented:
I added some documentation and removed the "optional" flag from 'dependency-submission'.

> Could you also add some explanation in a comment as to the change to running gradlew instead of running gradle-build-action ?

I presume this is targeted at new users of GitHub + Gradle, so I don't think it makes sense to document this as a change. It's natural for folks to run some sort of "setup" step followed by calling the build tool directly: having `gradle-build-action` do the execution was an anomaly IMHO.

I _did_ add a section showing what is required in the usual case that a user doesn't have the Gradle wrapper configured for their project. It looks a bit verbose, but it's complete.

jonjanego commented on ci/gradle.yml:
```diff
@@ -13,22 +13,55 @@ on:
+        java-version: '17'
+        distribution: 'temurin'
+
+    # Genereates and submits a dependency graph, enabling Dependabot Alerts for all project dependencies.
```
typo: "Generates"

bigdaz replied:
Fixed.

jonjanego reviewed (changes requested):
small spelling mistake, otherwise lgtm

jonjanego reviewed (approved):
looks good
